Assets/_BForBoss/_Core/Scripts/Managers/WaveManager.cs

```diff
@@ -7,54 +7,67 @@ namespace BForBoss
 {
     public class WaveManager : MonoBehaviour
     {
-        [Title("Components")]
-        [SerializeField] private EnemySpawnerManager _enemySpawnerManager;
-        
         [Title("Settings")] 
         [SerializeField, Tooltip("Number of enemies for the first wave")] private int _initialNumberOfEnemies = 10;
         [SerializeField, Tooltip("Cooldown time between waves")] private float _timeBetweenWaves = 2.5f;
         [SerializeField, Tooltip("Number of enemies per wave is the number of enemies from the previous wave multiplied by this multiplier")] private float _enemyAmountMultiplier = 1.2f;
 
-        private int _maxCountForCurrentWave = 0;
+
+        private EnemySpawnerManager _enemySpawnerManager;
+        private WaveModel _waveModel;
+        
         private int _spawnCount = 0;
-        private int _killCount = 0;
-        private int _waveNumber = 1;
 
-        public void Initialize(LifeCycleManager lifeCycleManager)
+        public void Initialize(LifeCycleManager lifeCycleManager, EnemySpawnerManager enemySpawnerManager, WaveModel waveModel)
         {
-            _maxCountForCurrentWave = _initialNumberOfEnemies;
-            _enemySpawnerManager.Initialize(lifeCycleManager, OnEnemySpawned, OnEnemyKilled);
+            _enemySpawnerManager = enemySpawnerManager;
+
+            _waveModel = waveModel;
+            _waveModel.waveNumber = 1;
+            _waveModel.maxEnemyCountforWave = _initialNumberOfEnemies;
+            _waveModel.killCount = 0;
+
+            _waveModel.OnEnemySpawned += OnEnemySpawned;
+            _waveModel.OnEnemyKilled += OnEnemyKilled;
+            
+            _enemySpawnerManager.Initialize(lifeCycleManager, waveModel);
         }
 
         public void Reset()
         {
-            _maxCountForCurrentWave = _initialNumberOfEnemies;
             _spawnCount = 0;
-            _killCount = 0;
-            _waveNumber = 1;
 
-            _enemySpawnerManager.Reset();
+            if (_waveModel != null)
+            {
+                _waveModel.ResetData();
+                _waveModel.OnEnemySpawned -= OnEnemySpawned;
+                _waveModel.OnEnemyKilled -= OnEnemyKilled;
```

Comment:
reset shouldn't unsub, 

you should only unsub on destroy

Reply:
Good catch, my bad